Utils/Dataflow/009_oracleConnector/dbConnection.py

```diff
@@ -121,10 +121,21 @@ def save_query_file(self, qname, src_filename, params=None):
             q = f.read().rstrip().rstrip(';')
             if isinstance(params, dict):
                 q = q % params
+            elif '%' in q.replace('%%', ''):
+                # Check for single '%' is done because a valid query without
+                # parameters (and their configuration) is possible.
+                sys.stderr.write("(WARN) No parameters were configured "
+                                 "for '%s' query. This can result in "
+                                 "'ORA-00911' error.\n" % qname)
```

Comment:
Not sure that I understand, why '%%' looks fine for you?

And '%' is still a valid symbol for LIKE statements... but you\`re right, it is hardly possible to know if this or that `%` is valid until one tries to compile query, so it is better to show the warning.

Reply:
> Not sure that I understand, why '%%' looks fine for you?

Let's take a look at different possible queries regarding this issue:

    1) SELECT a AS b FROM c;
    2) SELECT %(a)s AS b FROM c;
    3) SELECT a AS b%% FROM c;

In this function we perform `q % params` operation - it does nothing with 1 and 3, and changes 2 if the parameter 'a' was configured properly. The warning is about this operation of formatting, not the correctness of query in terms of Oracle - it mentions Oracle's problems because they can appear as a consequence of wrong formatting. If the configuration is missing, then the warning must appear only in case of 2: 1 is correct and needs no configuration, and 3 is also correct in terms of formatting. It's incorrect in terms of Oracle and will throw the same 'ORA-00911', but if we throw the same warning when 3 is encountered it would be misleading because the warning hints 'something with configuration', while actually the problem would be 'something with query'.

However...

> And '%' is still a valid symbol for LIKE statements... but you`re right, it is hardly possible to know if this or that '%' is valid until one tries to compile query, so it is better to show the warning.

... this is reasonable as well, so we can play on the safe side and just warn when any '%' is present. What do you think?

Reply:
> Let's take a look at different possible queries regarding this issue:
> 
> ```
> 1) SELECT a AS b FROM c;
> 2) SELECT %(a)s AS b FROM c;
> 3) SELECT a AS b%% FROM c;
> ```
> In this function we perform `q % params` operation - it does nothing with 1 and 3, and changes 2 if the parameter 'a' was configured properly.

Not exactly: after `q % params`  case 3 will become `SELECT a AS b% FROM c;`. But it is not to happen when no parameters specified (due to the check `if isinstance(params, dict)` before this statement).

Maybe add something to sound like "it looks like a parametric query, so here\`s the warning."?
```
(WARN) 'tasks': query seems to be parametric, but no query parameters were configured. May produce 'ORA-00911' error at compilation."
```

---

And, if (by chance) we want to be nice and give user a possibility to switch this warning off...
```
[queries]
 # absolute (or relative to config file) path to tasks query
 tasks: %__009_query_dir__%/prodsys2ES.sql
 # absolute (or relative to config file) path to datasets query
 datasets: %__009_query_dir__%/datasets.sql

+# Uncomment to switch off check for parameters configuration:
+# parametric_check: False
```
?

`True` by default, but if there are `LIKE` statements with proper `%` in the query and the warning is too annoying, it can be turned off.
Extra-complication, yes, but as it is obvious that the warning may appear when it is not welcome...

---

I also thought about more elaborate checks, like:
* step 1: is there '%'? If yes:
* step 2: try `query % {}`:
  * KeyError -- most likely it is a missed parameter;
  * ValueError, TypeError, passed -- most likely it has nothing to do with parameters.

How it looks and what produces this exceptions:
```
>>> # Looks like parametric query indeed
...
>>> query = 'SELECT %(a)s FROM b WHERE c;'
>>> query % {}
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
KeyError: 'a'
>>>
>>> # Not a parametric query
...
>>> query = 'SELECT %a FROM b WHERE c;'
>>> query % {}
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ValueError: unsupported format character 'a' (0x61) at index 8
>>>
>>> # Not a parametric query
...
>>> query = 'SELECT % FROM b WHERE c;'
>>> query % {}
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: float argument required, not dict
>>>
>>> # Doesn`t look much like a parametric query: parameter must be named
...
>>> query = 'SELECT %s FROM b WHERE c;'
>>> query % {}
'SELECT {} FROM b WHERE c;'
```

And, thinking about this "parameters must be named" -- it is also possible to look for `%(` instead of pure `%`.
In this case I think it is most likely that the warning won\`t appear when there were no sign of parametric query and thus this extra config parameter `parametric_check` is not needed (that much).

Reply:
> Maybe add something to sound like "it looks like a parametric query, so here`s the warning."

Makes sense. Done.

> And, thinking about this "parameters must be named" -- it is also possible to look for `%(` instead of pure `%`.

Good idea. Done.